File: gunhoflash/course-schedule.ts
function canFinish(numCourses: number, prerequisites: number[][]): boolean {
  const nPrerequisitesList: number[] = Array(numCourses).fill(0);
  const nextList: number[][] = Array(numCourses).fill(null).map(() => []);
  let nVisited = 0;

  for (const [nextCourse, prevCourse] of prerequisites) {
    nextList[prevCourse].push(nextCourse);
    ++nPrerequisitesList[nextCourse];
  }

  const visitingStack: number[] = [];
  for (let i = 0; i < numCourses; ++i) {
    if (nPrerequisitesList[i] === 0) {
      visitingStack.push(i);
    }
  }

  while (visitingStack.length) {
    ++nVisited;
    const i = visitingStack.pop()!;

    for (const nextCourse of nextList[i]) {
      if (--nPrerequisitesList[nextCourse] === 0) {
        visitingStack.push(nextCourse);
      }
    }
  }

  return nVisited === numCourses;
}

File: gunhoflash/course-schedule.py
from typing import List

class Solution:
  def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
    nPrerequisitesList = [0] * numCourses
    nextList = [[] for i in range(numCourses)]
    nVisited = numCourses

    for [nextCourse, prevCourse] in prerequisites:
      nextList[prevCourse].append(nextCourse)
      nPrerequisitesList[nextCourse] += 1

    visitingStack = []
    for i, n in enumerate(nPrerequisitesList):
      if n == 0:
        visitingStack.append(i)

    while len(visitingStack):
      nVisited -= 1
      i = visitingStack.pop()

      for nextCourse in nextList[i]:
        nPrerequisitesList[nextCourse] -= 1
        if nPrerequisitesList[nextCourse] == 0:
          visitingStack.append(nextCourse)

    return nVisited == 0
